### ISSUE TITLE
Type inference for array literals is incorrect

### ISSUE THREAD

**Issue by degory** [degory/ghul#597]
The element type of array literals is incorrectly inferred from the type of the first non-null element. The element type should be the most specific type that is assignable from every element. For example, given:

```
class Base is si
class Derived: Base is si

let o: Object;
let b: Base;
let d: Derived;
```  

then the type of `[b, b, d, d]` should be `Base[]`, the type of `[d, d, d, d]` should be `Derived[]`, and the type of both `[o, b, b, d]` and `[1234, b, b, d]` should be `Object[]`

**Comment by degory** [comment]
This has been partially fixed for some time - it uses the same heuristic as applied in other places where two or more types need to be joined into a single compatible type: see if the first type is assignable to the other, or vice versa and then pick the most derived. This is applied repeatedly to each element and the running best candidate type. It would benefit from proper least upper bound analysis though, which would allow inferring common ancestor types that do not appear as element types.

**Comment by degory** [comment]
Fixed by #1173
